[Sema]: ban the use of 'nonisolated' on actor declarations

jamieQ opened the pull request:
The combination doesn't make sense, so don't allow it.

Resolves: https://github.com/swiftlang/swift/issues/85174

jamieQ committed:
[Sema]: ban the use of 'nonisolated' on actor declarations

The combination doesn't make sense, so don't allow it.

jamieQ commented:
@swift-ci please test

jamieQ commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -7989,6 +7982,15 @@ void AttributeChecker::visitNonisolatedAttr(NonisolatedAttr *attr) {
+  if (auto *classDecl = dyn_cast<ClassDecl>(D);
+      classDecl && classDecl->isExplicitActor()) {
+    diagnoseAndRemoveAttr(attr, diag::invalid_decl_modifier, attr)
+        .warnUntilFutureSwiftVersion()
```
can we be more aggressive than this? i don't think this construction is likely that common...

ktoso replied:
I kicked off the source compat suite as well, we'd get some signal from it as well...

Generally it's better to not break source when it is low value like this

ktoso replied:
```suggestion
```

ktoso reviewed (approved):
Lgtm and thanks for including a test for distributed as well

jamieQ commented on test/decl/class/actor/basic.swift:
```diff
@@ -53,3 +53,5 @@ extension A2 {
   init(doesNotDelegate: ()) {} // expected-error {{designated initializer cannot be declared in an extension of 'A2'}}
 }
+
+nonisolated actor A3 {} // expected-warning {{'nonisolated' modifier cannot be applied to this declaration; this will be an error in a future Swift language mode}}{{1-13=}}
```
@ktoso just to confirm that i'm not off by one here... does the range here seem correct? i wasn't actually sure how to test the fixit 'for real'

ktoso commented on test/decl/protocol/special/DistributedActor.swift:
```diff
@@ -65,6 +65,8 @@ distributed actor D5: P1 {
   // expected-note@-1{{non-distributed instance method 'dist()'}}
 }
 
+nonisolated distributed actor D6 {} // expected-warning {{'nonisolated' modifier cannot be applied to this declaration; this will be an error in a future Swift language mode}}{{1-13=}}
```
```suggestion
nonisolated distributed actor D6 {} // expected-error {{'nonisolated' modifier cannot be applied to this declaration}}{{1-13=}}
```

ktoso reviewed (approved):
Approved and updating it to be a hard error after all; hope we can get away with it 🤞

ktoso committed:
Make 'nonisolated actor' a hard error

We've searched and can't find any usage of this in the wild, so hope we can get away with banning this outright.

ktoso commented:
Ok yeah let's do that with the actual error mode, thanks Pavel 👍
